New spinlocks in C++ w/tests

MagnusS opened the pull request:
TTAS spinlock implementation using C++ atomics. Implements C++11 named requirements BasicLockable, Lockable and TimedLockable.

Removes scoped_spinlock as we can now use `std::lock_guard` instead.

Updates all spinlock_t users to use the new class.

Includes unit test (w/multithreading) and integration test to check that the delays work.

This also fixes the alignment issue seen in https://github.com/includeos/IncludeOS/pull/2251#issuecomment-2305819667, as we're no longer aligning a `typedef` inside a class. I didn't add back `alignas` to the class as I'm not sure if it speeds anything up + it is Intel only. I suggest we consider adding it back if we have a concrete usecase we can benchmark.

I also removed the empty lock/unlock stubs that were added when SMP=off so now spinlocks always behave the same. If this leads to deadlocks anywhere we can fix it in the relevant code. This didn't happen in any of the tests I ran.

alfreb reviewed (approved):
We should move this to its own header, but besides that it looks excellent. I'm assuming all the tests are passing. If you prefer to merge as is, go ahead, but we'll then have to work on breaking apart the smp_utils later.

MagnusS commented on test/kernel/unit/spinlocks.cpp:
```diff
@@ -0,0 +1,179 @@
+  EXPECT(val == 2);
+}
+
+CASE("multithreaded - extended")
```
This runs as a unit test so I didn't want it to take too long or fail on less powerful machines. If I remove the locks I get the wrong value, so it seems to be enough to relatively reliably detect if the lock is broken (at least on my machine).

It could be interesting to run multiple cores in the integration test, but then we need SMP to work again first! :)

fwsGonzo commented on api/smp_utils:
```diff
@@ -20,38 +20,85 @@
+      // requires a write.
+      while (lock_.exchange(true, std::memory_order_acquire)) {
+        while (lock_.load(std::memory_order_relaxed)) {
+          _mm_pause();
```
`_mm_pause()` is purely x86 intrinsic, and won't work on arm where you have `__yield()` instead. It's basically a hint to the internal CPU scheduler, to aid hyperthreading.

MagnusS replied:
Ah, yes - then I guess we still have to include an architecture check here.

MagnusS replied:
I'm thinking about ways to implement this check. To try to avoid platform-specific checks outside the architecture specific include files, perhaps we could introduce an `os::Arch::cpu_relax` for each platform? Here: https://github.com/includeos/IncludeOS/blob/master/api/arch/x86_64.hpp#L19

We could then implement it as `_mm_pause` for x86 and skip the aarch64 implementation for now (so it would give a build error later) or just add it as an untested call to `__yield`.

What do you think @alfreb @fwsGonzo ?

alfreb replied:
That makes sense to me if the semantics of yield and pause are similar enough that they can use the same argument.

alfreb replied:
On the other hand - we don't actively support arm yet, and this lock should work just fine on arm as well without the intrinsic, so for me it's fine to do this later as well.

MagnusS commented:
Thanks for the reviews. I've updated to use `os::Arch::cpu_relax` instead of `_mm_pause` directly, which should allow us to easily add support for other architectures later. I also added a timeout for unittests and an EXCEPTS case I missed. I'll merge this for now -- then we can split it in .hpp files later -- and also decide what to do with minimal_barrier_t.
